### PR TITLE
VirusTotal - Updated the trusted vendors feature and added private API

### PR DESCRIPTION
**Pull request by guykeller**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16344

## Description
Fixed the integration's new scoring mechanism according to the desired functionality.
Also added the same functionality to the private API integration.

## Does it break backward compatibility?
   - No

## Must have
- [x] Documentation (with link to it)
- [x] Code Review

### CODE REVIEW

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 1]
```diff
@@ -52,14 +52,25 @@ configuration:
...
   type: 0
   required: false
+- display: Preferred Vendors List. List of vendors which are considered more trustworthy,
+    comma separated.
+  name: preferredVendors
+  defaultvalue: ""
+  type: 12
+  required: false
+- display: 'Preferred Vendor Threshold. The minimal number of highly trusted vendors
+    needed to consider a domain, IP, URL or file malicious. '
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal-Private_API.yml]:
```suggestion
    needed to consider a domain, IP, URL or file as malicious. '
```

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 2]
```diff
@@ -52,14 +52,25 @@ configuration:
...
+  defaultvalue: ""
+  type: 0
+  required: false
 script:
   script: |-
     ''' IMPORTS '''
     import requests
     import json
     import os
-    import time
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal-Private_API.yml]: why?

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 3]
```diff
@@ -69,15 +80,13 @@ script:
         del os.environ['http_proxy']
         del os.environ['https_proxy']
 
-    # Define utf8 as default encoding
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal-Private_API.yml]: why?

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 4]
```diff
@@ -87,6 +96,33 @@ script:
     ''' HELPER FUNCTIONS '''
 
 
+    def is_enough_preferred_vendors(scan_results):
+        if not (PREFERRED_VENDORS and PREFERRED_VENDORS_THRESHOLD):
+            return True
+        if PREFERRED_VENDORS and not PREFERRED_VENDORS_THRESHOLD:
+            return_error("Error: If you entered PREFERRED_VENDORS you must also enter PREFERRED_VENDORS_THRESHOLD")
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal-Private_API.yml]:
```suggestion
            return_error("Error: If you entered Preferred Vendors List you must also enter Preferred Vendor Threshold")
```

File: Integrations/integration-VirusTotal.yml [hunk 5]
```diff
@@ -99,7 +99,7 @@ script:
             return true;
         }
         if (PREFERRED_VENDORS && !(PREFERRED_VENDORS_THRESHOLD)) {
-            return("Error: If you entered Preferred Vendors List you must also enter Preferred Vendor Threshold")
+            return("Error: If you entered PREFERRED_VENDORS you must also enter PREFERRED_VENDORS_THRESHOLD")
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal.yml]:
```suggestion
            return("Error: If you entered Preferred Vendors List you must also enter Preferred Vendor Threshold")
```

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 6]
```diff
@@ -1589,7 +1630,6 @@ script:
       This call can be used in conjuction with the file searching call in order to
       download samples that match a given set of criteria.
   runonce: false
+releaseNotes: "Adding malicious attribute based on preferred vendors on File and URL"
```
Comment (ronykoz) [inline Integrations/integration-VirusTotal-Private_API.yml]: Add your RN to the old ones, they are still relevant

**Comment by glicht** [comment]
Can we add some tests?

**Comment by guykeller** [comment]
Tests were created and will now be added

**Comment by ronykoz** [comment]
@guykeller i think we should create a separate test for the preferred vendors scenario, as it might create confusion in the future in this test.
you can take a look at the `Panorama` tests which uses the `instance_names` feature(you should also update the PR in the content-test-conf repo)

**Comment by guykeller** [comment]
On it

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 7]
```diff
@@ -87,6 +101,33 @@ script:
...
+
+        for vendor_name in vendors_scans_dict:
+            cur_vendor_scan = vendors_scans_dict[vendor_name]
+            vendor_name_in_lowercase = vendor_name.lower()
+
+            if vendor_name_in_lowercase in list_of_preferred_vendors:
+                if "detected" in cur_vendor_scan and cur_vendor_scan["detected"]:
+                    counter += 1
+
+        return int(PREFERRED_VENDORS_THRESHOLD) <= counter
```
Comment (BenJoParadise) [inline Integrations/integration-VirusTotal-Private_API.yml]: Check during `test-instance`: `int(PREFERRED_VENDORS_THRESHOLD)` does not throw exception.

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 8]
```diff
@@ -87,6 +101,33 @@ script:
...
+            return_error("Error: If you entered Preferred Vendors you must also enter Preferred Vendors Threshold")
+        if "scans" not in scan_results:
+            return False
+
+        counter = 0
+
+        vendors_scans_dict = scan_results["scans"]
+        list_of_preferred_vendors = PREFERRED_VENDORS.split(',')
+
+        for i in range(len(list_of_preferred_vendors)):
```
Comment (BenJoParadise) [inline Integrations/integration-VirusTotal-Private_API.yml]: Use comprehension.
`.strip()` each element to handle case of `, ` or similar separator.

File: Integrations/integration-VirusTotal-Private_API.yml [hunk 9]
```diff
@@ -87,6 +101,33 @@ script:
...
+            list_of_preferred_vendors[i] = list_of_preferred_vendors[i].lower()
+
+        for vendor_name in vendors_scans_dict:
+            cur_vendor_scan = vendors_scans_dict[vendor_name]
+            vendor_name_in_lowercase = vendor_name.lower()
+
+            if vendor_name_in_lowercase in list_of_preferred_vendors:
+                if "detected" in cur_vendor_scan and cur_vendor_scan["detected"]:
```
Comment (BenJoParadise) [inline Integrations/integration-VirusTotal-Private_API.yml]: use `filter()` built-in function for lists, use `len()`.

File: TestPlaybooks/playbook-VirusTotalPrivateAPI_preferred_vendors_test.yml [hunk 10]
```diff
@@ -0,0 +1,197 @@
+id: virusTotalPrivateAPI-test-preferred-vendors
+version: 8
```
Comment (BenJoParadise) [inline TestPlaybooks/playbook-VirusTotalPrivateAPI_preferred_vendors_test.yml]:
```suggestion
version: -1
```

File: TestPlaybooks/playbook-VirusTotal_preferred_vendors_test.yml [hunk 11]
```diff
@@ -0,0 +1,254 @@
+id: virusTotaI-test-preferred-vendors
+version: 9
```
Comment (BenJoParadise) [inline TestPlaybooks/playbook-VirusTotal_preferred_vendors_test.yml]:
```suggestion
version: -1
```

File: TestPlaybooks/playbook-virusTotal-test.yml [hunk 12]
```diff
@@ -1,5 +1,5 @@
 id: virusTotal-test-playbook
-version: -1
+version: 3
```
Comment (BenJoParadise) [inline TestPlaybooks/playbook-virusTotal-test.yml]:
```suggestion
version: -1
```

File: TestPlaybooks/playbook-virusTotalPrivateAPI-test.yml [hunk 13]
```diff
@@ -1,67 +1,38 @@
 id: virusTotalPrivateAPI-test-playbook
-version: -1
+version: 9
```
Comment (BenJoParadise) [inline TestPlaybooks/playbook-virusTotalPrivateAPI-test.yml]:
```suggestion
version: -1
```

**Comment by yardensade** [comment]
Please merge it to master ASAP!!!

**Pull request merged**